app/controllers/discourse_data_explorer/query_controller.rb

```diff
@@ -227,6 +229,22 @@ def run
 
     private
 
+    def rate_limit_query_runs!
+      return if !is_api? && !is_user_api?
+
+      RateLimiter.new(
+        nil,
+        "api-query-run-10-sec",
+        GlobalSetting.max_data_explorer_api_reqs_per_10_seconds,
+        10.seconds,
+      ).performed!
+    rescue RateLimiter::LimitExceeded => e
+      if GlobalSetting.max_data_explorer_api_req_mode.include?("warn")
+        Discourse.warn("Query run 10 second rate limit exceeded")
```

Comment:
I wonder if we interpolate the query ID into this for extra data? You can pass an additional hash to `Discourse.warn`:

https://github.com/discourse/discourse/blob/3a9a534881135ac8928266b1d02bc679f2b49b31/app/models/optimized_image.rb#L347-L353

Reply:
Good idea! I saw it accepts a hash, but was on the fence since the query that triggers the rate limit is just "the straw that breaks the camel's back", but can probably still be useful for advocates. 🙂